src/store/normalize-data.js

```diff
@@ -85,13 +76,35 @@ const formatData = data => {
     if (data.schema_id) {
       state.id = data.schema_id;
     }
-    data.nodes.forEach(addNode(state));
-    data.edges.forEach(addEdge(state));
+
+    // add nodes and edges to state while building a node dependencies matrix for toposort
+    const nodeDeps = {};
+    const nodeIdsMap = {};
+    for (const node of data.nodes) {
+      addNode(state)(node);
+      nodeDeps[node.id] = [];
+      nodeIdsMap[node.id] = node;
+    }
+    for (const edge of data.edges) {
+      addEdge(state)(edge);
+      nodeDeps[edge.source].push(edge.target);
```

Comment:
As far as I can see, we are not doing that artificial node insertion, which may bring us some problems.

Let's imagine the following graph:

![mermaid-diagram-20200417104328](https://user-images.githubusercontent.com/44967953/79555976-52c13700-8098-11ea-858d-523336d8e7d2.png)

If I understand that correctly, Batching toposort will put `Raw 0` and `Int 0` into the same group (as they don't depend on each other). Then the order in which `Raw` and `Int` layers will be presented only depends on the order in which they appear in that toposort group (which is unreliable). So if I flip the branches on the graph, `Int` layer will be presented first, which is incorrect to me.

Gordon's suggestion was to add an artificial dependency node `Raw-to-Int` as in
![mermaid-diagram-20200417105148](https://user-images.githubusercontent.com/44967953/79556822-7fc21980-8099-11ea-9751-c0618334217a.png)
Which _depends_ on all `Raw` layer nodes and _is a dependency_ for all `Int` layer nodes. In that case toposort will put `Raw 0` and `Int 0` into different groups and `Raw` layer will go before `Int` layer.

I'm happy to discuss it over Zoom after the standup if necessary.